### ISSUE TITLE
Rate limiting

### ISSUE THREAD

**Issue by programcsharp** [wootwoot1234/Superhost-Tools#38]
Has anybody had issues with rate limiting? I've cribbed together some code based on this excellent project for my own purposes - messing around with some dynamic pricing rules -- but after doing a few (like 10 or 20) rate updates on a single listing via /v2/calendars/(listing)/(from)/(to), I get 503's every call after that until I wait for some period.

It's tough because the property I'm working on has a complex structure with like 50 different rates, and so the only way I can get through them is to do a few until 503, then wait, then try again later and do a few more.

I know this isn't really related to this project, but I was wondering if you have any tips on this -- really running into a wall here.

- Is there a known safe sleep period or rate for updating the pricing? 
- Or is there another api that allows bulking together a set of rates for one property into a single call instead having to spam a bunch of calls, one per date?

**Comment by wootwoot1234** [comment]
You might see if you can reduce the number of price updates you send.  For example, you could group days that are next to each other and have the same price or you could keep track of the current prices and not send updates for dates whos prices haven't changed.

**Comment by programcsharp** [comment]
Yep, I already do that where I can, but for stuff like weekends priced higher it's impossible -- you've got 52 * 2 groups minimum to do because you've got to do each one separately. I suppose you could do one background set for the entire year normal night price and then 52 sets for the weekends and that'd save some.

I just watched what Airbnb does on their multi calendar and they hit the `/api/v2/batch` endpoint and pass junk like:

```{"_transaction":false,"operations":[{"query":{"_format":"host_calendar_detailed"},"body":{"listing_id":1234567,"operations":[{"daily_price":284,"dates":["2019-04-05:2019-04-07"],"availability":"available","demand_based_pricing_overridden":true}]},"method":"POST","path":"/calendar_operations"}]}```

Looks like that's the same format as `/v2/calendars/(listing)/(from)/(to)`. Any experience with that? I bet it's way more efficient than doing a bunch of single calls... if it works.

**Comment by wootwoot1234** [comment]
I've used the `/api/v2/batch` to get information but not to post it.  I'm very interested in hearing more if you get it to work.  I think you're headed in the right direction.

**Comment by programcsharp** [comment]
Opened up a Chrome window to my Airbnb account and watched the XHR in inspector. 🤐

**Comment by wootwoot1234** [comment]
This might help.

method: `POST`
endpoint: `https://api.airbnb.com/v2/calendar_operations?_allow_dates_overlap=true`
x-airbnb-api-key: `915pw2pnf4h1aiguhph5gc5b2` (not sure if this matters)
data: `{"operations":[{"dates":["2018-03-13","2018-03-15","2018-03-19:2018-03-21"],"daily_price":111}],"listing_id":"XXXXXXXXX","method":"UPDATE"}`

Looks like you could probably send an array of `operations` all with different dates and prices.  I haven't played with it yet but you might be able to update your whole calendar with one call to the api.  Let me know how it goes.

**Comment by wootwoot1234** [comment]
@programcsharp did you have a chance to play try out that endpoint?  I interested because I think I'll use it as well and would also like to close this issue if you found a solution.

**Comment by programcsharp** [comment]
Finally got a chance to play with this... works a treat! Way faster than hitting the endpoint repeatedly to do multiple updates. I did basically what you have in your example (including all the overall junk like X-Airbnb-Api-Key, X-Airbnb-Oauth-Token, etc)...

The operations I filled with single dates if it is one night or a : separated range, just like you've got there.

Thanks for the pointers!
